x/auth/ante/sigverify.go

```diff
@@ -0,0 +1,350 @@
+	"github.com/cosmos/cosmos-sdk/x/auth/types"
+	"github.com/tendermint/tendermint/crypto"
+	"github.com/tendermint/tendermint/crypto/ed25519"
+	"github.com/tendermint/tendermint/crypto/multisig"
+	"github.com/tendermint/tendermint/crypto/secp256k1"
+)
+
+var (
+	// simulation signature values used to estimate gas consumption
+	simSecp256k1Pubkey secp256k1.PubKeySecp256k1
+	simSecp256k1Sig    [64]byte
+)
+
+func init() {
+	// This decodes a valid hex string into a sepc256k1Pubkey for use in transaction simulation
+	bz, _ := hex.DecodeString("035AD6810A47F073553FF30D2FCC7E0D3B1C0B74B61A1AAA2582344037151E143A")
+	copy(simSecp256k1Pubkey[:], bz)
+}
+
+// SignatureVerificationGasConsumer is the type of function that is used to both consume gas when verifying signatures
+// and also to accept or reject different types of PubKey's. This is where apps can define their own PubKey
+type SignatureVerificationGasConsumer = func(meter sdk.GasMeter, sig []byte, pubkey crypto.PubKey, params types.Params) error
+
+// Consume parameter-defined amount of gas for each signature according to the passed-in SignatureVerificationGasConsumer function
+// before calling the next AnteHandler
+type SigGasConsumeDecorator struct {
+	ak             keeper.AccountKeeper
+	sigGasConsumer SignatureVerificationGasConsumer
+}
+
+func NewSigGasConsumeDecorator(ak keeper.AccountKeeper, sigGasConsumer SignatureVerificationGasConsumer) SigGasConsumeDecorator {
+	return SigGasConsumeDecorator{
+		ak:             ak,
+		sigGasConsumer: sigGasConsumer,
+	}
+}
+
+func (sgcd SigGasConsumeDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate bool, next sdk.AnteHandler) (newCtx sdk.Context, err error) {
+	stdTx, ok := tx.(types.StdTx)
+	if !ok {
+		return ctx, errs.Wrap(errs.ErrTxDecode, "Tx must be a StdTx")
+	}
+
+	params := sgcd.ak.GetParams(ctx)
+
+	stdSigs := stdTx.GetSignatures()
+
+	// stdSigs contains the sequence number, account number, and signatures.
+	// When simulating, this would just be a 0-length slice.
+	signerAddrs := stdTx.GetSigners()
+
+	for i, sig := range stdSigs {
+		pubKey := sig.PubKey
+		if pubKey == nil {
+			signerAcc, err := GetSignerAcc(ctx, sgcd.ak, signerAddrs[i])
+			if err != nil {
+				return ctx, err
+			}
+			pubKey = signerAcc.GetPubKey()
+		}
+
+		if simulate {
+			// Simulated txs should not contain a signature and are not required to
+			// contain a pubkey, so we must account for tx size of including a
+			// StdSignature (Amino encoding) and simulate gas consumption
+			// (assuming a SECP256k1 simulation key).
+			consumeSimSigGas(ctx.GasMeter(), pubKey, sig, params)
+		}
+		err = sgcd.sigGasConsumer(ctx.GasMeter(), sig.Signature, pubKey, params)
+		if err != nil {
+			return ctx, err
+		}
+	}
+
+	return next(ctx, tx, simulate)
+}
+
+// Verify all signatures for tx and return error if any are invalid
+// increment sequence of each signer and set updated account back in store
+// Call next AnteHandler
+type SigVerificationDecorator struct {
```

Comment:
Need to figure out how to define the right type of interfaces to make this more general than just working with `auth.StdTx`

Reply:
How about the following interface:

```go
type SigVeriabiableTx struct {
  GetSignatures() []byte
  GetSigners() []sdk.AccAddress
  GetSignBytes(acc Account, chainID string) []byte
}
```

Reply:
I think that will work. Some decorators also need Pubkeys so i can add the method:

```go
GetPubKeys() []crypto.PubKey
```

Any StdSignature that has nil Pubkey will be ommitted from pubkey slice

Reply:
The problem is that the signature verification logic assumes a lot about the structure of the signature (signatures, account numbers, sequences, pubkeys, etc.)

I could turn everything into a method in an interface, but that seems messy.
I could also just assume that the signatures being returned are `StdSignature` structs, but that seems too restrictive

Reply:
Yes, it should be. Was wrong in earlier comment